pkg/controlplane/reconcilers/lease_test.go

```diff
@@ -22,63 +22,84 @@ https://github.com/openshift/origin/blob/bb340c5dd5ff72718be86fb194dedc0faed7f4c
 */
 
 import (
+	"reflect"
+	"sort"
 	"testing"
+	"time"
 
+	"github.com/google/uuid"
 	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/apitesting"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/storage"
+	etcd3testing "k8s.io/apiserver/pkg/storage/etcd3/testing"
+	"k8s.io/apiserver/pkg/storage/storagebackend/factory"
 	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/kubernetes/pkg/apis/core"
 	netutils "k8s.io/utils/net"
 )
 
-type fakeLeases struct {
-	keys map[string]bool
-}
+func init() {
+	var scheme = runtime.NewScheme()
 
-var _ Leases = &fakeLeases{}
+	metav1.AddToGroupVersion(scheme, metav1.SchemeGroupVersion)
+	utilruntime.Must(core.AddToScheme(scheme))
+	utilruntime.Must(v1.AddToScheme(scheme))
+	utilruntime.Must(scheme.SetVersionPriority(v1.SchemeGroupVersion))
 
-func newFakeLeases() *fakeLeases {
-	return &fakeLeases{make(map[string]bool)}
+	codecs = serializer.NewCodecFactory(scheme)
 }
 
-func (f *fakeLeases) ListLeases() ([]string, error) {
-	res := make([]string, 0, len(f.keys))
-	for ip := range f.keys {
-		res = append(res, ip)
-	}
-	return res, nil
-}
+var codecs serializer.CodecFactory
 
-func (f *fakeLeases) UpdateLease(ip string) error {
-	f.keys[ip] = true
-	return nil
+type fakeLeases struct {
+	storageLeases
 }
 
-func (f *fakeLeases) RemoveLease(ip string) error {
-	delete(f.keys, ip)
-	return nil
-}
+var _ Leases = &fakeLeases{}
 
-func (f *fakeLeases) SetKeys(keys []string) {
-	for _, ip := range keys {
-		f.keys[ip] = false
+func newFakeLeases(t *testing.T, s storage.Interface) *fakeLeases {
+	// TODO make NewLeases sanitize paths, though it is only used in one place
+	// pkg/controlplane/instance.go:268:       masterLeases, err := reconcilers.NewLeases(config, "/masterleases/", ttl)
+	// add a random prefix so we can reuse the etcd instance for subtests
+	base := "/" + uuid.New().String() + "/masterleases/"
```

Comment:
Smart - I was looking at the test for 3 minutes trying to understand why the tests are independent before I realized this line...

Reply:
this is how we do in the integration test :)